core/src/main/java/novoda/lib/sqliteprovider/util/DatabaseStructure.java

```diff
@@ -0,0 +1,226 @@
+    private static final String PRAGMA_TABLE_INFO = "PRAGMA table_info('%1$s');";
+    private static final String PRGAMA_INDEX_LIST = "PRAGMA index_list('%1$s');";
+    private static final String PRGAMA_INDEX_INFO = "PRAGMA index_info('%1$s');";
+    private static final String COLUMN_PRIMARY_KEY_INDEX = "pk";
+    private static final String COLUMN_NAME = "name";
+    private static final String COLUMN_TYPE = "type";
+
+    private static final List<String> DEFAULT_TABLES = Collections.singletonList("android_metadata");
+
+    private final SQLiteDatabase database;
+
+    private List<String> createdTables;
+    private Map<String, List<String>> foreignTables = new TreeMap<>();
+    private Map<String, List<Constraint>> uniqueConstraints = new TreeMap<>();
+
+    public DatabaseStructure(SQLiteDatabase database) {
+        this.database = database;
+    }
+
+    public List<String> tables() {
+        final Cursor tablesCursor = database.rawQuery(SELECT_TABLES_NAME, null);
+        createdTables = new ArrayList<>(tablesCursor.getCount());
+        parseTables(tablesCursor);
+        tablesCursor.close();
+        return Collections.unmodifiableList(createdTables);
+    }
+
+    private void parseTables(Cursor tablesCursor) {
+        while (tablesCursor.moveToNext()) {
+            addCreatedTableIfNotDefault(tablesCursor.getString(0));
+        }
+    }
+
+    private void addCreatedTableIfNotDefault(String tableName) {
+        if (!DEFAULT_TABLES.contains(tableName)) {
+            createdTables.add(tableName);
+        }
+    }
+
+    public Map<String, String> projectionMap(String parent, String... foreignTables) {
+        Map<String, String> projection = new TreeMap<>();
+        projection.put("_id", parent + "._id AS _id");
+        projection.putAll(projectionFor(parent, columns(parent)));
+        projection.putAll(projectionFor(foreignTables));
+        return Collections.unmodifiableMap(projection);
+    }
+
+    private static Map<String, String> projectionFor(String table, Map<String, SQLiteType> fields) {
+        Map<String, String> projection = new TreeMap<>();
+        for (Map.Entry<String, SQLiteType> entry : fields.entrySet()) {
+            projection.put(table + "_" + entry.getKey(), table + "." + entry.getKey() + " AS "
+                    + table + "_" + entry.getKey());
+        }
+        return projection;
+    }
+
+    private Map<String, String> projectionFor(String... foreignTables) {
+        Map<String, String> projection = new TreeMap<>();
+        for (String ft : foreignTables) {
+            projection.putAll(projectionFor(ft, columns(ft)));
+        }
+        return projection;
+    }
+
+    public List<String> foreignTables(String table) {
+        foreignTables.put(table, new ArrayList<String>(5));
+        final Cursor columnsCursor = queryTableColumnsFor(table);
+        List<String> allTables = tables();
```

Comment:
This is an example of a place that could cause issues if someone inlined that `createdTables` field.